Add swupd plugin

albertomurillo opened the pull request:
##### ISSUE TYPE
 - New Module Pull Request

##### COMPONENT NAME
swupd.py

##### ANSIBLE VERSION
```
ansible 2.3.0 (clearlinux 234556d37a) last updated 2017/01/30 20:01:49 (GMT +000)
  config file =
  configured module search path = Default w/o overrides
```

##### SUMMARY
Add support for ClearLinux update manager (swupd)

Commit:
Add swupd plugin

Add support for ClearLinux update manager (swupd)

Signed-off-by: Alberto Murillo Silva <alberto.murillo.silva@intel.com>

jtyr commented on lib/ansible/modules/packaging/os/swupd.py:
```diff
@@ -0,0 +1,288 @@
+module: swupd
+short_description: Manages bundles with I(swupd)
+description:
+    - "Manages bundles with the I(swupd) update manager, which is used by the
```
Standard YAML indentation is 2 spaces.

albertomurillo replied:
Done

jtyr commented on lib/ansible/modules/packaging/os/swupd.py:
```diff
@@ -0,0 +1,288 @@
+      Clear Linux Project for Intel Architecture"
+version_added: "2.3"
+author:
+    - "Alberto Murillo (@albertomurillo)"
```
This string doesn't have to be quoted.

albertomurillo replied:
Done

jtyr commented on lib/ansible/modules/packaging/os/swupd.py:
```diff
@@ -0,0 +1,288 @@
+        choices: ["present", "absent"]
+    update:
+        description:
+            - Updates the OS to the latest version
```
Please add dot at the end. The same applies to the descriptions bellow.

albertomurillo replied:
Done

jtyr commented on lib/ansible/modules/packaging/os/swupd.py:
```diff
@@ -0,0 +1,288 @@
+        default: null
+    contenturl:
+        description:
+            - url for content file download
```
the `url` should be `URL`. The same bellow.

albertomurillo replied:
Done

jtyr commented on lib/ansible/modules/packaging/os/swupd.py:
```diff
@@ -0,0 +1,288 @@
+    cmd = "%s %s" % (SWUPD_CMD, command)
+
+    if p["format"]:
+        cmd += " --format=%s" % (p["format"])
```
The the formatting argument is single value, it doesn't have to be in braces:

```
cmd += " --format=%s" % p["format"]
```

It improves the readability slightly. The same should apply across the whole code.

albertomurillo replied:
Done

jtyr commented on lib/ansible/modules/packaging/os/swupd.py:
```diff
@@ -0,0 +1,288 @@
+    """The main function."""
+    module = AnsibleModule(
+        argument_spec=dict(
+            contenturl=dict(default=None, type='str'),
```
The `default=None` can be removed. And I think the `type=str` can be removed as well as it's the default type.

albertomurillo replied:
I removed the default=None but kept type='str' as it is easier to know that the type is.

jtyr commented on lib/ansible/modules/packaging/os/swupd.py:
```diff
@@ -0,0 +1,288 @@
+    if p['state'] == "present":
+        swupd_install(module, p['name'])
+
+    if p['state'] == "absent":
```
This can be as `elif`.

albertomurillo replied:
All calls to `swupd_*` functions result in `module.exit_json()` or `module.fail_json()` but that is not very clear in the main() function so I added `elif` as suggested

jtyr reviewed (changes requested):
Please also use the `flake8` tool to check your code before submitting.

jtyr commented on lib/ansible/modules/packaging/os/swupd.py:
```diff
@@ -0,0 +1,286 @@
+      - Name of the (C)bundle to install or remove.
+    required: false
+    default: null
+    aliases: ['bundle']
```
You don't have to quote YAML strings. Quote only `yes` or `no` to prevent it to turn into `true` or `false`. Please also unify the quotes to either single or double quotes.

albertomurillo replied:
done

jtyr commented on lib/ansible/modules/packaging/os/swupd.py:
```diff
@@ -0,0 +1,286 @@
+module: swupd
+short_description: Manages bundles with M(swupd).
+description:
+  - "Manages bundles with the M(swupd) update manager, which is used by the
```
This doesn't have to be quoted.

albertomurillo replied:
done

jtyr commented on lib/ansible/modules/packaging/os/swupd.py:
```diff
@@ -0,0 +1,286 @@
+version_added: "2.3"
+author: Alberto Murillo (@albertomurillo)
+options:
+  name:
```
The list of options should be in alphabetical order.

albertomurillo replied:
done

jtyr commented on lib/ansible/modules/packaging/os/swupd.py:
```diff
@@ -0,0 +1,286 @@
+options:
+  name:
+    description:
+      - Name of the (C)bundle to install or remove.
```
Shouldn't the `(C)bundle` be just `bundle`? The same bellow.

albertomurillo replied:
`bundle` is an alias of `name` so it is an option too. Actually I was thinking on setting the option to bundle and the alias to name since ClearLinux's swupd doesn't manage packages, just bundles.

jtyr replied:
If it's an alias (effectively an option name) then it should be written like `I(bundle)`.

albertomurillo replied:
done

jtyr commented on lib/ansible/modules/packaging/os/swupd.py:
```diff
@@ -0,0 +1,286 @@
+    default: null
+  url:
+    description:
+      - Overrides both C(contenturl) and C(versionurl).
```
`C()` should be used only for values. For names of options, `I()` should be used instead. Please correct it everywhere in the `DOCUMENTATION` block.

albertomurillo replied:
From [developing modules checklist](http://docs.ansible.com/ansible/dev_guide/developing_modules_checklist.html)
```
Ensure that you make use of U() for urls, C() for files and options, I() for params, M() for modules.
```

I'll ask in IRC for clarification

jtyr replied:
Yeah. That's a bit confusing as it's not clearwhat's `options` and what's `params`. If you check other modules you will see that the option names are decorated with `I()` (e.g. `I(state)`) and option values are decorated with `C()` (e.g. `C(present)`).

albertomurillo replied:
So maybe we need a fix on the documentation then.

jtyr replied:
Here it's: https://github.com/ansible/ansible/pull/20931

albertomurillo replied:
done

jtyr commented on lib/ansible/modules/packaging/os/swupd.py:
```diff
@@ -0,0 +1,286 @@
+
+    p = module.params
+
+    global SWUPD_CMD
```
What about to turn all the above functions into an object and then set this global variable as the object variable? It would look more elegant. See the comment bellow.

jtyr replied:
Not interested to do the object-oriented approach?

albertomurillo replied:
Ok, I created the swupd class, see the next update

jtyr commented on lib/ansible/modules/packaging/os/swupd.py:
```diff
@@ -0,0 +1,286 @@
+    if not SWUPD_CMD:
+        module.fail_json(msg="Could not find swupd.")
+
+    if p['update']:
```
What about to create shortcuts for all these `p['...']` variables? It would be easier to read the code then:

```
# Create the object
swupd = SwUpd(module)

# Create shortcuts
update = p['update']
verify = p['verify']
state = p['state']
name = p['name']

# Trigger action
if update:
  swupd.update()
elif verify:
  swupd.verify()
elif state == "present":
  swupd.install(name)
elif state == "absent":
  swupd.remove(name)
```

If you would implement the object approach, then you can create more shortcuts in the `__init__()` method. For example for `p["format"]`, `p["manifest"]`, `p["contenturl"]` and `p["versionurl"]` and then use `self.format`, `self.manifest`, `self.contenturl` and `self.versionurl` instead.

albertomurillo replied:
I implemented the shortcuts despite being a few more lines it does make the main() function more readable

jtyr commented on lib/ansible/modules/packaging/os/swupd.py:
```diff
@@ -0,0 +1,286 @@
+    elif p['state'] == "absent":
+        swupd_remove(module, p['name'])
+
+    module.fail_json(msg="Unkown Error")
```
Normally, the success should be reported from the `main()` function. The fuctions above (or the object methods if you decide to implement them) should only return values which are used to compose the `exit_json()`.

albertomurillo replied:
So following the `fail-fast` method, I think is better to call `fail_json` as soon as an error is encountered, and also in places where you do not expect it like the end of a function that should never get to the end.

albertomurillo replied:
kind of the same for `exit_json` as it allows the following code to be easier to read without so much indentation and `elifs`. I have seen this approach by various other `packages/os` modules

albertomurillo replied:
also we dont want to do anything else after one of the calls is made.

albertomurillo replied:
Ok I am returning fail and exit from the main() function now except for `check_mode` which does not need to perform anything on the host.

jtyr commented on lib/ansible/modules/packaging/os/swupd.py:
```diff
@@ -0,0 +1,286 @@
+    if module.check_mode:
+        module.exit_json(changed=_is_bundle_installed(module, bundle))
+
+    if not _is_bundle_installed(module, bundle):
```
This could be `elif`.

albertomurillo replied:
Not really need to be elif since the return is explicit there.

jtyr commented on lib/ansible/modules/packaging/os/swupd.py:
```diff
@@ -0,0 +1,286 @@
+    if rc == 0:
+        module.exit_json(changed=True, msg="Bundle %s installed" % bundle, stdout=stdout, stderr=stderr)
+
+    if rc == 18:
```
This could be `elif`.

albertomurillo replied:
Not really need to be elif since the return is explicit there.

albertomurillo replied:
same as above

jtyr commented on lib/ansible/modules/packaging/os/swupd.py:
```diff
@@ -0,0 +1,286 @@
+    if module.check_mode:
+        module.exit_json(changed=not _is_bundle_installed(module, bundle))
+
+    if _is_bundle_installed(module, bundle):
```
This could be `elif`.

albertomurillo replied:
Not really need to be elif since the return is explicit there.

jtyr replied:
The same like above - return (exit) at the end?

albertomurillo replied:
Its better to `fail-fast` and avoid unnecesary `elif` and indentation clutter

jtyr commented on lib/ansible/modules/packaging/os/swupd.py:
```diff
@@ -0,0 +1,286 @@
+    if rc == 0:
+        return True
+
+    if rc == 1:
```
This could be `elif`.

albertomurillo replied:
Not really need to be elif since the return is explicit there.

jtyr replied:
True. I kind of like to store the return value into a var and return at the end of the function/method than from the middle of the code.

albertomurillo replied:
I think this style is more readable, there is no need for another variable for such a simple function

jtyr commented on lib/ansible/modules/packaging/os/swupd.py:
```diff
@@ -0,0 +1,286 @@
+from ansible.module_utils.basic import AnsibleModule
+
+
+ANSIBLE_METADATA = {'status': ['preview'],
```
This could have nicer formatting:

```
ANSIBLE_METADATA = {
    'status': ['preview'],
    'supported_by': 'community',
    'version': '1.0'
}
```

albertomurillo replied:
While I agree that format looks better, the [documentation](http://docs.ansible.com/ansible/dev_guide/developing_modules_checklist.html) and practically all other modules (except for those missing ANSIBLE_METADATA) use the current format

jtyr replied:
All modules have the same format because it was added automatically only recently. Nothings can stop you to improve that formatting.

albertomurillo replied:
done

jtyr commented on lib/ansible/modules/packaging/os/swupd.py:
```diff
@@ -0,0 +1,324 @@
+            module.fail_json(msg="Could not find swupd.")
+
+        # Initialize parameters
+        self.contenturl = module.params["contenturl"]
```
Just looking at this, it could be simplified by using something like this:

```
# Create shortcuts
for key in module.params.keys():
  setattr(self, key, module.params[key])
```

albertomurillo replied:
I like it. Done

mattclay commented:
CI is failing due to a python 2.4 syntax error:

```
2017-02-02 03:26:53 Compiling ./lib/ansible/modules/packaging/os/swupd.py ...
2017-02-02 03:26:53   File "./lib/ansible/modules/packaging/os/swupd.py", line 115
2017-02-02 03:26:53     class Swupd():
2017-02-02 03:26:53                 ^
2017-02-02 03:26:53 SyntaxError: invalid syntax
```

jtyr commented:
The Python 2.4 error should be fixed by changing the class line to either:

```
class Swupd:
```

or

```
class Swupd(object):
```

I kind of like more the first one as it's shorter ;o)

jtyr commented on lib/ansible/modules/packaging/os/swupd.py:
```diff
@@ -0,0 +1,317 @@
+# along with Ansible.  If not, see <http://www.gnu.org/licenses/>.
+
+import os
+from ansible.module_utils.basic import AnsibleModule
```
Although it's again my believes, the project [code style guidelines](https://github.com/ansible/proposals/issues/50) require to have the imports under the `DOCUMENTATION`, `EXAMPLES` and `RESULTS`.

albertomurillo replied:
@jtyr This has already been taken care of, I think you were looking at an old version of the PR

jtyr replied:
Hehe. I have written that comment days ago but forgot to submit it ;o)

albertomurillo commented:
@jtyr Do you have any more issues you want me to fix on this PR before I can have your +1 ?

jtyr commented:
I did not test the module so I think I should not flag it for shipment. But from the code review point of view, I'm giving +1.

gundalow commented on lib/ansible/modules/packaging/os/swupd.py:
```diff
@@ -0,0 +1,316 @@
+    verify: yes
+    manifest: 12920
+'''
+
```
RETURNS are not documented

albertomurillo replied:
Fixed

jtyr replied:
As per the [documentation](http://docs.ansible.com/ansible/common_return_values.html), the `stdout` and the `stderr` are common values to all modules and therefore should not be documented in the module itself.

bcoca replied:
they are not common to 'all' modules, but they are already documented as they are common to many modules

jtyr replied:
So do the common values have to be documented in the module or not?

bcoca replied:
they don't need to, but can be as long as they are acurate

gundalow commented on lib/ansible/modules/packaging/os/swupd.py:
```diff
@@ -0,0 +1,316 @@
+DOCUMENTATION = '''
+---
+module: swupd
+short_description: Manages bundles with M(swupd).
```
Manages bundles on Clear Linux systems

gundalow replied:
I'm not sure if we can also use the words "packages" or "updates" in their to make it a little more obvious.

albertomurillo replied:
Fixed

gundalow commented on lib/ansible/modules/packaging/os/swupd.py:
```diff
@@ -0,0 +1,316 @@
+module: swupd
+short_description: Manages bundles with M(swupd).
+description:
+  - Manages bundles with the M(swupd) update manager, which is used by the
```
No need for the `M(..)` here as that will just generate a link to the same page.

albertomurillo replied:
fixed

gundalow commented on lib/ansible/modules/packaging/os/swupd.py:
```diff
@@ -0,0 +1,316 @@
+
+- name: Check integrity of filesystem
+  swupd:
+    verify: yes
```
Does this return anything, such as where there are validation errors?
An of how to access that could be useful (see `register:` and `debug`)

albertomurillo replied:
This module returns stderr and stdout when swupd is actually run.
Cases when swupd does not run includes installing a bundle which is already installed (file exist in /usr/share/clear/bundles/bundlename) and the module returns stdout and stderr as empty strings

gundalow commented on lib/ansible/modules/packaging/os/swupd.py:
```diff
@@ -0,0 +1,316 @@
+        swupd.failed = True
+
+    if swupd.failed:
+        module.fail_json(msg="Unkown Error")
```
Typo of "unknown".
Can we give the user any clues, for example asking them to check a log on the remote system?

albertomurillo replied:
Fixed

gundalow commented on lib/ansible/modules/packaging/os/swupd.py:
```diff
@@ -0,0 +1,316 @@
+
+
+class Swupd(object):
+    FILES_NOT_MATCH = "files did not match"
```
What happens if the system us running in a different local?

albertomurillo replied:
swupd currently does not support different local so it would print "files did not match"

Commit:
swupd: Document RETURN values

Signed-off-by: Alberto Murillo Silva <alberto.murillo.silva@intel.com>

Commit:
swupd: Add a meaningful description for failures

Signed-off-by: Alberto Murillo Silva <alberto.murillo.silva@intel.com>

Commit:
swupd: Clarify documentation

Provide more information about the options supported by the module
since some concepts are new for people not used to clearlinux/swupd

Signed-off-by: Alberto Murillo Silva <alberto.murillo.silva@intel.com>
